### ISSUE TITLE
Loading optimizer tensors from Checkpoint leads to inefficient organization of data on GPU?

### ISSUE THREAD

**Issue by sjscotti** [facebookresearch/ParlAI#4237]
Hi
I modified `torch_agent.py` to use the shampoo optimizer at https://github.com/google-research/google-research/tree/master/scalable_shampoo/pytorch to train a BB2 model.  The optimizer ran when doing a fresh start for training (or restarting just after vacuuming a checkpoint).   However, when I did a restart from a checkpoint without vacuuming, I got an error that some of the tensors needed for the optimizer were on the CPU and not on the GPU (BTW, this optimizer keeps a lot more state info than the optimizers built into parlai).  So I wrote some code to move the tensors to the GPU right after the optimization data was read in from the checkpoint.  Anyway, when I got the optimizer tensors loaded on the GPU and started with the training again, it seemed to take much longer for a training step.   I have a limited GPU memory so a lot of my training depends on using "virtual" GPU memory borrowed from the CPU.    So I rewrote the code that moves the checkpoint optimizer tensors to the GPU to have it run at the first call to the `step` routine instead of when the optimizer data is read from the checkpoint.  Now the training steps take about the same amount of time as when doing a fresh start.  So does it make sense that having the tensors specific to the optimizer routine (such as momentum) as the last ones created on the GPU is more efficient?  Does that also mean that for a restart, the optimizer specific tensors shouldn't be read in until they are needed for first optimizer step?

**Comment by sjscotti** [comment]
As an alternative, for a restart from a checkpoint, there a way to load the optimizer state into CPU memory?  I think I can figure out how to move it to GPU memory when needed by the optimizer.

**Comment by sjscotti** [comment]
Though I think this is still a good enhancement suggestion, as a workaround before I restart from the checkpoint, I do a 
`parlai vacuum --model-file xxx.checkpoint` (where `xxx.checkpoint` is the model file name)
so that the checkpoint optimizer state is discarded.  The penalty is that I need to do a warmup when I restart from the checkpoint.  Keeping the warmup short compared to the length of the restart run limits the computational cost.
I'll close out this issue.

**Comment by stephenroller** [comment]
We do always load the full state dict (including the optimizer) into CPU from disk, since we can't guarantee parameters are on the same GPU as they were during training:

https://github.com/facebookresearch/ParlAI/blob/fc33db25adda708ccdf45d24a899b5b50ab8fc75/parlai/core/torch_agent.py#L2075

I believe load state dict copies the weights of the optimizer to wherever the optimizer is living, regardless of whether the tensors are originally on CPU:

https://github.com/facebookresearch/ParlAI/blob/fc33db25adda708ccdf45d24a899b5b50ab8fc75/parlai/core/torch_agent.py#L1113

This is handled within pytorch's optimizer load state dict:
https://github.com/pytorch/pytorch/blob/06838ce8b16b2cc2f9e903f3ebdd46659a0e66bb/torch/optim/optimizer.py#L161
